Generate interface for opaque types in kotlin plugin to enable JVM test fakes.

vaibhav-1207 opened the pull request:
Generate interface for corresponding generated class of opaque types in kotlin backend to enable JVM test fakes.

Interface generation is customized via a argument passed to backend attribute directive `#[diplomat::attr(kotlin, generate_mocking_interface]`.

Currently, the `generate_mocking_interface` attribute is enabled only for opaque types and kotlin backend, other types or backend will throw a lowering error.

Manishearth reviewed (approved):
I'd want @emarteca  or @jcrist1 to also approve this, since they maintain the kotlin backend

jcrist1 commented:
Will have a look on the weekend, but a quick question @vaibhav-1207. Can you create an example of how to use this new interface?

emarteca commented:
Thanks! I will also look in the next couple days. I agree with @jcrist1 that you should include an example of how this is meant to be used (and I'll loop in the relevant people on my side too)

vaibhav-1207 commented:
Hi, I have updated the PR with an example. I've added the `generate_mocking_interface` attribute and updated the corresponding Kotlin class. Please review it when you have time. Thanks!

jcrist1 commented on tool/src/kotlin/mod.rs:
```diff
@@ -2514,4 +2527,56 @@ mod test {
+            let type_name = opaque_def.name.to_string();
+            // test that we can render and that it doesn't panic
+            let (_, result) =
+                ty_gen_cx.gen_opaque_def(opaque_def, &type_name, "dev.gigapixel", "somelib", true);
```
This is definitely my fault for introducing it in the first place 😅 but could you get rid of my personal handle?
```suggestion
                ty_gen_cx.gen_opaque_def(opaque_def, &type_name, "dev.diplomattest", "somelib", true);
```

jcrist1 reviewed (changes requested):
This looks fine, but could you introduce the mocked interface in a test? e.g. create a test that has a mocked opaque and a regular opaque and assert it produces the same values or something along those lines?

jcrist1 commented:
> This looks fine, but could you introduce the mocked interface in a test? e.g. create a test that has a mocked opaque and a regular opaque and assert it produces the same values or something along those lines?

Also @vaibhav-1207 sorry for taking so long to get to the actual code review. It fell a bit off my radar cause I didn't see it in PR reviews tab in GH.

vaibhav-1207 commented:
I've made the following changes -
1. Added feature test mocking for the generated Kotlin backend interface.
2. Removed personal handle references from the Kotlin plugin's tests.

However, I'm running into issue getting the Kotlin tests to execute. When I try `cargo make test-kotlin-feature` or `gradle test`, I get the following error:
`
|FAILURE:` Build failed with an exception.

What went wrong:
Task 'test' not found in root project 'somelib'.

Try:
Run gradle tasks to get a list of available tasks. Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
`

I'm bit stuck trying to debug the error and am relying on the workflow tests once approved. Please let me know if you have any insights about this error. 

Please review the changes when you have time. Thanks!

jcrist1 commented:
@vaibhav-1207 I should have time to look at this on Thursday. Thanks for the example implementation.

jcrist1 commented:
@vaibhav-1207 I'm able to run the tests locally but I get the following issues:
```
> Task :compileKotlin
e: file:///Users/cristina/workspace/rust/review/diplomat/feature_tests/kotlin/somelib/src/main/kotlin/dev/diplomattest/somelib/AttrOpaque1.kt:52:9 'method' hides member of supertype 'AttrOpaque1Interface' and needs 'override' modifier
e: file:///Users/cristina/workspace/rust/review/diplomat/feature_tests/kotlin/somelib/src/main/kotlin/dev/diplomattest/somelib/AttrOpaque1.kt:58:9 'abirenamed' hides member of supertype 'AttrOpaque1Interface' and needs 'override' modifier
e: file:///Users/cristina/workspace/rust/review/diplomat/feature_tests/kotlin/somelib/src/main/kotlin/dev/diplomattest/somelib/AttrOpaque1.kt:64:9 'useUnnamespaced' hides member of supertype 'AttrOpaque1Interface' and needs 'override' modifier
e: file:///Users/cristina/workspace/rust/review/diplomat/feature_tests/kotlin/somelib/src/main/kotlin/dev/diplomattest/somelib/AttrOpaque1.kt:70:9 'useNamespaced' hides member of supertype 'AttrOpaque1Interface' and needs 'override' modifier
```

It looks like the interface methods need to be explicitly overriden. 

As to your problem, are you able to load the project into IntelliJ (Community edition)? I think that might help troublshoot.

jcrist1 commented:
Also what versions of gradle and java are you using?

jcrist1 commented:
@vaibhav-1207 have you been able to look into the overrides?

vaibhav-1207 commented:
@jcrist1 thanks for reminding and sorry that I missed your reply because of team summit. I'll tackle this sometime this week.

vaibhav-1207 commented:
Thank you @jcrist1.  I was indeed using an outdated gradle version (4.4.1), which caused compatibility issues with a dependency, and it was fixed after updating the gradle to latest stable version.

I've also fixed the above error by updating the Jinja template to add override modifier for the interface methods, and have verified the fix locally using `cargo make test-kotlin-feature`. PTAL!

jcrist1 reviewed (approved):
Thanks for digging into this! LGTM to me now, other than the override coming before the comment, but I think that's just cosmetic. If you can fix it quickly that would be nice, but I don't want to hold this up much more. I tested the kotlin compile locally. @Manishearth can we trigger the ci for this?

vaibhav-1207 commented on example/kotlin/somelib/src/main/kotlin/dev/diplomattest/somelib/FixedDecimal.kt:
```diff
@@ -48,7 +47,7 @@ class FixedDecimal internal constructor (
         }
     }
-    
+    override
```
Thanks for the catch. I've updated the method template, and that should resolve it.

vaibhav-1207 commented on tool/templates/kotlin/Method.kt.jinja:
```diff
@@ -3,7 +3,11 @@
+{%- if add_override_specifier_for_opaque_self_methods %}
+override {{ declaration }} {
+{%- else %}
 {{ declaration }} {
```
Thanks, inlined the chnages.

Manishearth commented:
Triggered. I'll wait for @emarteca to give the go-ahead for merging (she can also trigger CI as needed)

vaibhav-1207 commented:
I see 2 failures - 
1. `rustfmt of kotlin/mod.rs` - I've fixed the fmt with the latest change
2. `test failure of kotlin::test::test_opaque_gen_with_mocking_interface` - I am not able to reproduce this error on my local setup. I ran `cargo test` & `cargo insta test` and all the tests are passing, so I am not sure why the tests are failing here.

jcrist1 commented on tool/src/kotlin/snapshots/diplomat_tool__kotlin__test__opaque_gen_with_mocking_interface.snap:
```diff
@@ -0,0 +1,57 @@
+    fun MyOpaqueStruct_get_byte(): FFIUint8
+    fun MyOpaqueStruct_get_string_wrapper(in1: Int): Int
+}
+internal interface MyOpaqueStructInterface {
+}
```
@vaibhav-1207  this seems to be an issue. I would expect this to have the trait definition (and of course the overrides below). I don't know immediately why it's not there.

jcrist1 replied:
ahh n/m I see that this opaque doesn't actually have any non-static functions

vaibhav-1207 replied:
I've added a change to generate interfaces only when the Rust opaque struct has any non-static functions.

vaibhav-1207 commented on tool/src/kotlin/snapshots/diplomat_tool__kotlin__test__opaque_gen_with_mocking_interface.snap:
```diff
@@ -0,0 +1,57 @@
+import com.sun.jna.Pointer
+import com.sun.jna.Structure
+
+
```
yeah, a rerun might help!

I'll also work on a change to only generate the interface when an opaque has any non-static functions, but that shouldn't hold up this PR.

vaibhav-1207 commented:
@jcrist1 Thanks for pointing this out! I rebased with updated `main` branch, reproduced the failure locally, and updated the snapshot. However, I'm not super familiar with the git command-line ops and seem to have messed something up. While the snapshot is fixed, all the commits are now duplicated - I think I made some mistake while merging and unifying diverging branches on my fork!

I am not sure if this current state works for you. If not, I can look into cleaning up this mess!

vaibhav-1207 commented:
Hey @jcrist1, can we merge this PR if there are no other remaining issues?
